examples/binding_affinity_prediction/main.py

```diff
@@ -23,80 +25,168 @@ def run_a_train_epoch(args, epoch, model, data_loader,
+            prediction = model(bg)
         loss = loss_criterion(prediction, (labels - args['train_mean']) / args['train_std'])
         epoch_loss += loss.data.item() * len(indices)
         optimizer.zero_grad()
         loss.backward()
         optimizer.step()
         train_meter.update(prediction, labels)
     avg_loss = epoch_loss / len(data_loader.dataset)
+    # if (args['num_epochs'] - epoch) <= 6: # print only the last 5 epochs
     total_scores = {metric: train_meter.compute_metric(metric, 'mean')
                     for metric in args['metrics']}
     msg = 'epoch {:d}/{:d}, training | loss {:.4f}'.format(
         epoch + 1, args['num_epochs'], avg_loss)
     msg = update_msg_from_scores(msg, total_scores)
     print(msg)
+    return total_scores
 
 def run_an_eval_epoch(args, model, data_loader):
     model.eval()
     eval_meter = Meter(args['train_mean'], args['train_std'])
     with torch.no_grad():
         for batch_id, batch_data in enumerate(data_loader):
             indices, ligand_mols, protein_mols, bg, labels = batch_data
-            labels, bg = labels.to(args['device']), bg.to(args['device'])
-            prediction = model(bg)
+            labels = labels.to(args['device'])
+            if args['model'] == 'PotentialNet':
+                bigraph_canonical, knn_graph = bg # unpack
+                bigraph_canonical = bigraph_canonical.to(args['device'])
+                knn_graph = knn_graph.to(args['device'])
+                prediction = model(bigraph_canonical, knn_graph)
+            elif args['model'] == 'ACNN':
+                bg = bg.to(args['device'])
+                prediction = model(bg)
             eval_meter.update(prediction, labels)
     total_scores = {metric: eval_meter.compute_metric(metric, 'mean')
                     for metric in args['metrics']}
     return total_scores
 
 def main(args):
+    torch.multiprocessing.set_sharing_strategy('file_system')
     args['device'] = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
     set_random_seed(args['random_seed'])
 
-    dataset, train_set, test_set = load_dataset(args)
+    train_set, val_set, test_set = load_dataset(args)
     args['train_mean'] = train_set.labels_mean.to(args['device'])
     args['train_std'] = train_set.labels_std.to(args['device'])
+
+    if args['test_on_core']:   # always test on core set
+        test_loader_dict = {'subset': 'core', 'frac_train': 0, 'frac_val': 0, 'frac_test': 1}
+        args.update(test_loader_dict)
+        _, _, test_set = load_dataset(args)
+
     train_loader = DataLoader(dataset=train_set,
                               batch_size=args['batch_size'],
-                              shuffle=False,
-                              collate_fn=collate)
+                              shuffle=args['shuffle'],
+                              collate_fn=collate,
+                              pin_memory= True,
+                              num_workers=8)
```

Comment:
It might be better to allow users to specify the number of workers to use as this depends on the resources then can access.